source/SH3/camera/camera.cpp

```diff
@@ -0,0 +1,131 @@
+/** @file
+ *
+ *  Implementation of camera.hpp
+ *
+ *  @copyright 2017 Palm Studios
+ *
+ *  @date 1-9-2017
+ *
+ *  @author Jesse Buhagiar
+ */
+#include <cmath>
+#include "SH3/camera/camera.hpp"
+#include "glm/gtc/matrix_transform.hpp"
+#include <algorithm>
+#include <boost/algorithm/clamp.hpp>
+
+static constexpr glm::vec3 worldUp = glm::vec3(0, 1, 0); /**< Which axis is considered 'world up'. */
+
+sh3::camera::Camera::Camera(const glm::vec3& pos, angle<float> fov, float aspect, float zNear, float zFar)
+: camPos(pos), camFov(fov), aRatio(aspect), camNear(zNear), camFar(zFar),
+  camFront(glm::vec3(0.0f, 0.0f, -1.0f)), camUp(glm::vec3(0.0f, 1.0f, 0.0f)),
+  camPitch(angle<float>::FromDegrees(-90.0f)),
+  camYaw(angle<float>::FromDegrees(0.0f)),
+  camMode(MODE::FIRST_PERSON)
+{
+    RecalculatePerspective();
+}
+
+void sh3::camera::Camera::RecalculatePerspective()
+{
+    glm::vec3 front;
+    front.x = std::cos(camPitch.AsRadians()) * std::cos(camYaw.AsRadians());
+    front.y = std::sin(camPitch.AsRadians());
+    front.z = std::cos(camPitch.AsRadians()) * std::sin(camYaw.AsRadians());
+    camFront = glm::normalize(front);
+
+    camRight = glm::normalize(glm::cross(camFront, worldUp));
+    camUp = glm::normalize(glm::cross(camRight, camFront));
+
+    vMatrix = glm::lookAt(camPos, camPos + camFront, camUp);
+    pMatrix = glm::perspective(camFov.AsRadians(), aRatio, camNear, camFar); // Set our perspective matrix
+}
+
+glm::mat4 sh3::camera::Camera::GetProjectionMatrix() const
+{
+    return pMatrix * glm::lookAt(camPos, camPos + camFront, camUp);
```

Comment:
Any reason this isn't `pMatrix * vMatrix`?
Also, can't we already do the multiplication in `RecalculatePerspective()` and store it in `pMatrix`? If so, then this method can return a `const glm::mat4&` fine.

Reply:
Ahh very true. I've stored the product of the two matrices in pmatrix. I'm merely going to comment out vmatrix for now, as I have a feeling we may encounter a time where we'll need access to the view matrix/send it to the shader (though for the most part, the view and model matrix are actually combined, the view-model matrix.)